### ISSUE TITLE
incorrectly handle insertText (macOS)

### ISSUE THREAD

**Issue by ksqsf** [rust-windowing/winit#3925]
### Description

# Description

Currently, the implementation of `insertText` is insufficient to cover some input methods that can commit text without setting a preedit. (e.g. when the input method translates a certain key and directly commits the translation.)

Known affected downstream programs include Alacritty and Neovide.

# Examples

- A typical example is the full-width punctuation. If you hit the period key with the input method enabled, it will commit `。` to the program **without** ever setting a preedit.
- Another example is SKK-style Japanese input methods, with which, e.g. `a` directly commits `あ` without any preedit. Currently, both full-width punctuation and SKK Japanese input method doesn't work properly.

# Analysis

Upon inspection, I found that `insertText` will only commit text when there is a preedit. (Hence the description above.)

I removed the condition and it *almost* worked for me. However, it seems to be added in commit 3fd73848 for a good reason.

https://github.com/rust-windowing/winit/blob/7e819bb2ce0c5f3aa3acf04000e27444812ea239/src/platform_impl/apple/appkit/view.rs#L406-L411

# Notes

By "almost" above, I mean it's still not fully satisfactory.

- On startup, the IME is disabled and only gets enabled after a preedit is set once. That is, for full-width punctuation to work, one has to input something to enter the IME state. For SKK, the situation could get a bit problematic. This should also be fixed. (I think this probably belongs to a separate issue? It could be fixed by setting `Ground` to be the default. But I might be missing some details here.)
- Some input methods support toggle keys. For example, in SKK, q toggles Hiragana and Katagana input. If I removed the preedit check, whenever I type `q`, the input method can receive `q` (toggling katagana) and another `q` will also be forwarded to the application. 

Another thing is that I noticed the system input methods can commit full-width punctuation just fine. But to my surprise I didn't manage to find any difference in the logs. So I don't really understand what's going on.

### macOS version

```shell
ProductName:		macOS
ProductVersion:		15.0
BuildVersion:		24A335
```

### Winit version

0.30.3

**Comment by fredizzimo** [comment]
> On startup, the IME is disabled and only gets enabled after a preedit is set once.

I think this might additionally explain the following two bugs (just based on the symptoms, so I might be wrong)
* https://github.com/rust-windowing/winit/issues/3095
* https://github.com/rust-windowing/winit/issues/3342

**Comment by kchibisov** [comment]
>I removed the condition and it almost worked for me. However, it seems to be added in commit https://github.com/rust-windowing/winit/commit/3fd73848dd36cc051cc3a0bec305ba8b2a117145 for a good reason.

That's the reason you have `KeyboardInput` in the first place, without this line, you don't have it anymore, thus, said apps can not really function, since everything is a `Commit` regardless of what layout are you using, which is just wrong. So if anyone want to solve this, they should ensure that `KeyboardInput` is still _there_, if it's not, your solution doesn't work.

**Comment by lifei** [comment]
what's the status of this pr?

**Comment by ksqsf** [comment]
@lifei No progress. I don't have an idea yet.

I guess the correct way forward is to reinterpret system results as key events rather than `Ime::Committed` and such, similar to how Emacs handles it. To make that work, the client app must be aware whether it is expecting text or not, but that doesn't seem to be feasible currently. 

OTOH, if the client app is willing to indicate whether it wants keys or text by calling `set_ime_allowed`, this PR should work as-is: say in a GUI app, the user focuses a text entry and the widget lib calls `set_ime_allowed(true)`, and calls with `false` when it's unfocused.

**Comment by fredizzimo** [comment]
You can still deliver normal `KeyboardInput` events even when IME is allowed, but the same input can't generate both `KeyboardInput` and `Ime`.

**Comment by ksqsf** [comment]
@fredizzimo Yep, I'm aware of that. The current version behaves that way by heuristically deciding whether to send key events to IME.

I'll try to summarize the problem here:

1. There is no way to tell if a key *should* be handled by IME => we have to forward all keys to IME first, or we cannot support special IMEs. (SKK is an extreme example because it basically breaks all IME conventions.)
2. IME invokes certain ObjC methods in the app. However, it does not distinguish unhandled raw strings from converted strings. For example, the US keyboard layout will call `insertText`, in exactly the same way how a CJK IME would behave.

To winit, without a hint from the app itself, we cannot tell with certainty if the key event should be directly forwarded to the app, or it is the text from IME should be sent to the app. The root problem is due to winit distinguishing between the two, I'd say, which may make sense for Windows or Linux (I don't know) but definitely causes some problems for macOS.

**Comment by kchibisov** [comment]
I'm pretty sure it's possible to fix this, since the sequence should be generally different, but given that I'm not that interested myself in figuring out macOS nor apple allows me to use newer versions, I'm not that much interested myself.

The thing is that certain characters which don't work have likely a special pattern, since they are not being inserted in a normal way at all IIRC, so figuring that pattern out is how it should be done.

I know that maybe `unmarked` text may be used for that, from what I've seen in certain toolkits, but generally I don't see how it's impossible, since browsers somehow can deal with it, given that they implement W3C spec on macOS as well.

**Comment by kchibisov** [comment]
I also don't see any issue in e.g. looking into how firefox builds the sequence and how it deals with the IME input, or e.g. chrome. I'm sure they should have the same issue and probably solved it somehow.

**Comment by cathaysia** [comment]
I encounter this bugs when using alacritty:

https://github.com/user-attachments/assets/0fd4ec31-f777-4d8b-b964-55e8f1ae762f

step:

1. press `Ctrl+Shift+0` switch punctuation to chinese punctuation.
2. the IME try switch punctuation, and the terminal still output `0`.
3. and the IME cannot output chinese punctuation.

I tried having Codex analyze the cause, and eventually found it here.

I had never encountered a similar problem before when using kitty/wezterm, so I had codex analyze the differences between the two:

Ultimately, it came to the following conclusion; I wonder if it will be helpful in addressing this issue.

---

Kitty's processing order is roughly as follows:

macOS keyDown
    ↓
NSTextInputClient / interpretKeyEvents
    ↓
Rime decides: commit text, enter preedit, or swallow the shortcut
    ↓
Kitty decides whether to continue sending the original keystroke based on the IME result
    ↓
PTY / tmux

There are four key differences.

1. Kitty accepts "direct commits without preedit"

Chinese punctuation is often not:

    Preedit(".") → Commit("。")

Instead, Rime directly calls:

    insertText("。")

Without any preedit/marked text.

Kitty's insertText (https://github.com/kovidgoyal/kitty/blob/v0.46.2/glfw/cocoa_window.m#L1917) will unconditionally save this commit, even if there was no marked text before:

Press . → Rime commits directly. → Kitty accepts.

The problem with winit #3925 is exactly the opposite: it has long only recognized insertText when preedit exists, so it might discard ., leaving only the original ..

2. Kitty queries the IME first, then decides whether to send the key.

Kitty's keyDown (https://github.com/kovidgoyal/kitty/blob/v0.46.2/glfw/cocoa_window.m#L1257) first calls:

[self interpretKeyEvents:@[event]];

This gives Rime its first chance to process. Kitty then checks:

- Whether the IME has produced a commit

- Whether preedit exists or has just ended

- Whether insertText has cleared marked text

- Whether the IME has processed the key but not generated text

If the IME is processing and has not generated plain text, Kitty will return directly without sending the original key to PTY. This is the key to preventing the error message "Rime switch successful, but a 0 was entered at the same time".
